ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,291 @@
+                for j in i + 1..x.len() {
+                    basis_polynomial *= p[k] - x[j];
+                }
+            }
+            sum += basis_polynomial;
+        }
+        // compute "result_k = result_k * p(p_k)"
+        result[k] *= sum;
+    }
+}
+
+mod test {
+    use rand::{thread_rng, Rng};
+
+    use crate::{
+        ff::ec_prime_field::Fp25519,
+        protocol::ipa_prf::malicious_security::lagrange::{
+            compute_lagrange_base, compute_lagrange_denominator, generate_evaluation_points,
+            lagrange_evaluation, lagrange_evaluation_precomputed,
+        },
+        secret_sharing::SharedValue,
+    };
+
+    #[test]
+    fn lagrange_evaluation_fp25519() {
+        let mut rng = thread_rng();
+        let evaluation_points_size = rng.gen::<usize>() % 100;
+        let mut evaluation_points = vec![Fp25519::ONE; evaluation_points_size];
+        evaluation_points
+            .iter_mut()
+            .for_each(|x| *x = rng.gen::<Fp25519>());
+        let mut y_values = vec![Fp25519::ONE; evaluation_points_size];
+        y_values.iter_mut().for_each(|y| *y = rng.gen::<Fp25519>());
+        let mut evaluated_y = vec![Fp25519::ONE; evaluation_points_size];
+        lagrange_evaluation(
+            &evaluation_points,
+            &y_values,
+            &evaluation_points,
+            &mut evaluated_y,
+        );
+
+        // test evaluation at interpolation points
+        assert_eq!(evaluated_y, y_values);
+
+        // sample random polynomial in monomial form of degree `evaluation_points_size`
+        let mut polynomial = vec![Fp25519::ONE; evaluation_points_size];
+        polynomial
+            .iter_mut()
+            .for_each(|x| *x = rng.gen::<Fp25519>());
+        // add random point to evaluation points
+        evaluation_points.push(rng.gen::<Fp25519>());
+
+        // evaluate polynomial at evaluation_points
+        let mut y_values = vec![Fp25519::ZERO; evaluation_points_size + 1];
+        evaluation_points.iter().for_each(|x_value| {
+            let mut base = Fp25519::ONE;
+            polynomial.iter().for_each(|coefficient| {
+                y_values.iter_mut().for_each(|y| *y = *coefficient * base);
+                base *= *x_value
+            })
+        });
+
+        // lagrange evaluate at random point
+        evaluated_y[0] = Fp25519::ONE;
+        lagrange_evaluation(
+            &evaluation_points[0..evaluation_points_size],
+            &y_values[0..evaluation_points_size],
+            std::slice::from_ref(&evaluation_points[evaluation_points_size]),
+            std::slice::from_mut(&mut evaluated_y[0]),
+        );
+
+        // check equality
+        assert_eq!(y_values[evaluation_points_size], evaluated_y[0]);
+    }
+
+    #[test]
+    fn lagrange_precomputed_test() {
+        let mut rng = thread_rng();
+        let evaluation_points_size = rng.gen::<usize>() % 100;
+        let mut evaluation_points_random = vec![Fp25519::ONE; evaluation_points_size];
```

Comment:
these tests are large and hard to read. Consider moving repetitive code into helper function and using type system as suggested above.